{"metadata": {"language_info": {"name": "python", "version": "3.6.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# Extract SNPs from gentoype data\n\n\n## Overview\n\nThis notebook allows the extraction of individual SNPs from the UK Biobank genotyping data. \n<br>\nSNPs of interest can specified in the following ways:\n\n* A list of individual SNPs\n* Genomic coordinates\n\n* A text file containing an rsID for each SNP on a separate line, for example:\n\nrs683743847\n<br>\nrs73473437\n\n* A text file containing genomic coordinates, where the first column lists the chromosome, the following two columns specify the start and end coordinates and the last column provides an identifier for the region. Do not include column headers.\n\n    |Column1| Column2| Column3| Column4|\n    |-------|--------|--------|--------|\n    |2 | 30000000 | 35000000 | R| \n    |2 | 60000000 | 62000000 | R2\n    |X |10000000  | 20000000 | R3\n    \n     \n     \n\n\n\nIn order to extract SNPs, the plink binary files holding the genotype calls are downloaded to the instance and merged into a single binary file. This file is then used for SNP extraction. Although these operations could be performed for all files containing genotype calls, it is faster and less computationally intensive to specify chromosomes of interest first and download and merge only genotype call files relevant to the chromosomes of interest. \n\n", "metadata": {}}, {"cell_type": "markdown", "source": "# How to run this notebook\n\nFollow the steps below to run this Jupiter Notebook:\n\n* Click on the Tools menu and select \"JupiterLab\"\n* Click on the \"New JupiterLab\" button to start a JupiterLab instance.\n* Select a name and a project from the dropdown meny for your JupiterLab environment.\n* Select the priority for you JupiterLab environment; \"Normal\" of \"High\" is recommended.\n* Under \"Cluster Configuration\", select Single Node.\n* Set instance type and duration for you environment.If the chromosomes where SNPs of interest are known in advance and a small number of chromosomes files are  selected for download, an instance type with such as mem1_hdd1_v2_x8 or mem1_hdd1_v2_x4 should be sufficient. If all chromosome files are selected for download, instance type with more computational resources may be needed (mem1_hdd1_v2_x16 or above).\n* Click on \"Start Environment\"\n* You will see your environment go from \"Initialising\" to \"Launching\" and then \"Ready\". This may take some time depending on the priority selected; at busy times, it may be necessary to select high priority to avoid long initialising times. Once the environment is ready, click on \"Open\".\n* A JuiterLab session will open. On the left side of the screen, you will see a a \"DNA Nexus\" tab, allowing you to open notebooks directly from your project environment. If you have saved this notebook under you project environment, just double click to open it.\n* Press \"Ctrl\" + \"Enter\" to run code cells. An hourglass icon on the JupireLab tab in your browser indicates that the code is running. Please note that depending on number of chromosomes and SNPs and your instance type, code may take some time to run.", "metadata": {}}, {"cell_type": "markdown", "source": "## Install plink \n\nPlink is installed using conda and plink --help is called to verify the installation.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\n# install plink\nconda install -c bioconda plink\nplink --help | head", "metadata": {"trusted": true}, "execution_count": 1, "outputs": [{"name": "stdout", "text": "Collecting package metadata (current_repodata.json): ...working... done\nSolving environment: ...working... done\n\n## Package Plan ##\n\n  environment location: /opt/conda\n\n  added / updated specs:\n    - plink\n\n\nThe following packages will be downloaded:\n\n    package                    |            build\n    ---------------------------|-----------------\n    ca-certificates-2022.4.26  |       h06a4308_0         124 KB\n    plink-1.90b6.21            |       h516909a_0         6.8 MB  bioconda\n    ------------------------------------------------------------\n                                           Total:         7.0 MB\n\nThe following NEW packages will be INSTALLED:\n\n  plink              bioconda/linux-64::plink-1.90b6.21-h516909a_0\n\nThe following packages will be UPDATED:\n\n  ca-certificates    conda-forge::ca-certificates-2021.10.~ --> pkgs/main::ca-certificates-2022.4.26-h06a4308_0\n\n\nProceed ([y]/n)? Invalid choice: plink --help | head\nProceed ([y]/n)? \n\nDownloading and Extracting Packages\nca-certificates-2022 | 124 KB    | ########## | 100% \nplink-1.90b6.21      | 6.8 MB    | ########## | 100% \nPreparing transaction: ...working... done\nVerifying transaction: ...working... done\nExecuting transaction: ...working... done\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Download genotyping files \n\nPlink files containing genotypes are downloaded to the local instance using the dx download command.\n\nIf the chromosome location of the SNPs is known in advance , it may be best to limit the number of downloaded files to files for the chromosomes needed using the options below.\n", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\n# Use the code below to download files for chromosomes of interest\n\n\n# list all available genotype calls files and places them in a file\n# The file all_snp_files.txt will appear in your environment (click on the folder icon on the left of your screen to view files in your environment)\ndx find data --name \"ukb22418_*.bed\" --path \"/Bulk/Genotype Results/Genotype calls/\" --delim ',' | awk -F, '{ print $4 }' |awk -F'/' '{print $5}' | awk -F'.' '{print $1}'  > all_snp_files.txt\n\n\n# The code below provides an example on how to download data for  chromosome 1, chromosome 4 and chromosome X\n# To modify, replace  the 1,4, and X in the followig expression '^ukb22418_c[14X]_' with your chromosome of interest\n# For example, if you are interested in SNPs on chromosomes 5,6,8, and 9, use: '^ukb22418_c[5689]_'\n# For double-digit chromosomes , e.g chromosome 12 use the following format: '^ukb22418_c[1][2]_'\n# For mixed double digit and single digit chromosomes, e,g, chromosome 8 and 12, use the following format: '^ukb22418_c[18][2]_'\n# To include all chromosomes modify the expression to '^ukb22418_c*_' \n\n\n\nfiles=$(grep '^ukb22418_c[14X]_' all_snp_files.txt) \necho $files \n\n\n\n\n# Download files containing chromosomes of interest\n\nfor i in $files; do\ndx download \"/Bulk/Genotype Results/Genotype calls/${i}*\"\ndone\n\n\n\n# The downloaded files will appear in your environment. There should be a bed, bim and fam file for each chromosome containing SNPs of interest.\n\n", "metadata": {"trusted": true}, "execution_count": 2, "outputs": [{"name": "stdout", "text": "ukb22418_cX_b0_v2 ukb22418_c4_b0_v2 ukb22418_c1_b0_v2\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Merge genotyping files\n\nGenotyping files containing data for chromosomes of interest are merged into a single binary format file, genotyping_merged.\nThe following files should appear in the file list:\n\ngenotyping_merged.bed\n<br>\ngenotyping_merged.bim\n<br>\ngenotyping_merged.fam\n\nPlease note code in this cell might take some time to run.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\n# Define files to merge. Please replace the '^ukb22418_c[14X]_' expression in the code below with the expression used to define chromosomes in the previous code chunk. \n\ngrep '^ukb22418_c[14X]_' all_snp_files.txt > files_to_merge.txt\ncat files_to_merge.txt\nplink --merge-list files_to_merge.txt --make-bed --out genotyping_merged", "metadata": {"trusted": true}, "execution_count": 3, "outputs": [{"name": "stdout", "text": "ukb22418_cX_b0_v2\nukb22418_c4_b0_v2\nukb22418_c1_b0_v2\nPLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to genotyping_merged.log.\nOptions in effect:\n  --make-bed\n  --merge-list files_to_merge.txt\n  --out genotyping_merged\n\n15544 MB RAM detected; reserving 7772 MB for main workspace.\nPerforming 3-pass merge (488377 people, 44408/129787 variants per pass).\nPass 1 complete.                              \nPass 2 complete.                              \nMerged fileset written to genotyping_merged-merge.bed +\ngenotyping_merged-merge.bim + genotyping_merged-merge.fam .\n129787 variants loaded from .bim file.\n488377 people (223429 males, 264742 females, 206 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to genotyping_merged.nosex .\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 488377 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.974043.\n129787 variants and 488377 people pass filters and QC.\nNote: No phenotypes present.\n--make-bed to genotyping_merged.bed + genotyping_merged.bim +\ngenotyping_merged.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Extarcting individual SNP IDs\n\nUse the code cell below if you want to extract data for a small number of SNPs.\nProvide a list of SNP rsIDs, separated by commas.\nYour results can be found in the snp_ind_plink_results.raw The file is uploaded to your project space and can be downloaded from there.\n", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n# replace the rsIDs in the code below with your SNPs; Always seperate each rsID with a commma.\nplink --bfile genotyping_merged --snps rs28659788,rs116587930 --recode A --out snp_ind_plink_results\ndx upload snp_ind_plink_results.raw", "metadata": {"trusted": true}, "execution_count": 4, "outputs": [{"name": "stdout", "text": "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to snp_ind_plink_results.log.\nOptions in effect:\n  --bfile genotyping_merged\n  --out snp_ind_plink_results\n  --recode A\n  --snps rs28659788,rs116587930\n\n15544 MB RAM detected; reserving 7772 MB for main workspace.\n129787 variants loaded from .bim file.\n488377 people (223429 males, 264742 females, 206 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to snp_ind_plink_results.nosex .\n--snps: 2 variants remaining.\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 488377 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.500514.\n2 variants and 488377 people pass filters and QC.\nNote: No phenotypes present.\n--recode A to snp_ind_plink_results.raw ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\nID                    file-GF2fpF0J8v5Zp4k71qkkf7yj\nClass                 file\nProject               project-G96qBfjJ8v5XFFfK7yxxx5y3\nFolder                /\nName                  snp_ind_plink_results.raw\nState                 closing\nVisibility            visible\nTypes                 -\nProperties            -\nTags                  -\nOutgoing links        -\nCreated               Wed Jul  6 11:08:28 2022\nCreated by            lora.boteva\n via the job          job-GF2fX58J8v5bYY7Z2JV8QQpx\nLast modified         Wed Jul  6 11:08:31 2022\nMedia type            \narchivalState         \"live\"\ncloudAccount          \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Extracting a list of SNPs provided in a text file\n\nUse the code cell below if you have a file listsing rsIDs of interest.\nPrepare a file containing rsIDs (one per line). Rename it to snps_list.txt and upload to your project folder.\nThe file is downloaded to the instance using the dx dowload command and used to extract SNPs. \nYour results can be found in the snp_list_plink_results.raw The file is uploaded to your project space and can be downloaded from there.\n", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\ndx download -f \"snps_list.txt\"\nplink --bfile genotyping_merged --extract snps_list.txt --recode A --out snp_list_plink_results\ndx upload snp_list_plink_results.raw\n", "metadata": {"trusted": true}, "execution_count": 5, "outputs": [{"name": "stdout", "text": "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to snp_list_plink_results.log.\nOptions in effect:\n  --bfile genotyping_merged\n  --extract snps_list.txt\n  --out snp_list_plink_results\n  --recode A\n\n15544 MB RAM detected; reserving 7772 MB for main workspace.\n129787 variants loaded from .bim file.\n488377 people (223429 males, 264742 females, 206 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to snp_list_plink_results.nosex .\n--extract: 10 variants remaining.\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 488377 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.887258.\n10 variants and 488377 people pass filters and QC.\nNote: No phenotypes present.\n--recode A to snp_list_plink_results.raw ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\nID                    file-GF2fpX8J8v5x9YVyKxzfy8zb\nClass                 file\nProject               project-G96qBfjJ8v5XFFfK7yxxx5y3\nFolder                /\nName                  snp_list_plink_results.raw\nState                 closing\nVisibility            visible\nTypes                 -\nProperties            -\nTags                  -\nOutgoing links        -\nCreated               Wed Jul  6 11:08:57 2022\nCreated by            lora.boteva\n via the job          job-GF2fX58J8v5bYY7Z2JV8QQpx\nLast modified         Wed Jul  6 11:09:00 2022\nMedia type            \narchivalState         \"live\"\ncloudAccount          \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Extracting SNPs from a genomic region\n\nThe code cell below allows SNPs to be exported for a defined genomic region. Change the --chr, --from-kb and --to-kb flags as required.\nYour results can be found in the snp_region_plink_results.raw The file is uploaded to your project space and can be downloaded from there.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nplink --bfile genotyping_merged --chr 1 --from-kb 0 --to-kb 30000000 --recode A --out snp_region_plink_results\ndx upload snp_region_plink_results.raw", "metadata": {"trusted": true}, "execution_count": 8, "outputs": [{"name": "stdout", "text": "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to snp_region_plink_results.log.\nOptions in effect:\n  --bfile genotyping_merged\n  --chr 1\n  --from-kb 0\n  --out snp_region_plink_results\n  --recode A\n  --to-kb 30000000\n\n15544 MB RAM detected; reserving 7772 MB for main workspace.\n63487 out of 129787 variants loaded from .bim file.\n488377 people (223429 males, 264742 females, 206 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to snp_region_plink_results.nosex .\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 488377 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.974083.\n63487 variants and 488377 people pass filters and QC.\nNote: No phenotypes present.\n--recode A to snp_region_plink_results.raw ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\nID                    file-GF2g4yQJ8v5ggVBBKxb86Y5X\nClass                 file\nProject               project-G96qBfjJ8v5XFFfK7yxxx5y3\nFolder                /\nName                  snp_region_plink_results.raw\nState                 closing\nVisibility            visible\nTypes                 -\nProperties            -\nTags                  -\nOutgoing links        -\nCreated               Wed Jul  6 11:31:06 2022\nCreated by            lora.boteva\n via the job          job-GF2fX58J8v5bYY7Z2JV8QQpx\nLast modified         Wed Jul  6 11:38:08 2022\nMedia type            \narchivalState         \"live\"\ncloudAccount          \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Extracting SNPs from a list of genomic regions\n\nThe code cell below allows you to filter SNPs using a file lsiting numerous genomic locations.\nPrepare a file containing the required genomic locations (one per line). Rename it to genomic_regions.txt and upload to your project folder.\nThe file is downloaded to the instance using the dx dowload command and used to extract SNPs.\nYour results can be found in the snp_region_list_plink_results.raw The file is uploaded to your project space and can be downloaded from there.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\ndx download -f \"genomic_regions.txt\"\n\nplink --bfile genotyping_merged --extract range genomic_regions.txt  --recode A --out snp_region_list_plink_results\n\ndx upload snp_region_list_plink_results.raw\n", "metadata": {"trusted": true}, "execution_count": 7, "outputs": [{"name": "stdout", "text": "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to snp_region_list_plink_results.log.\nOptions in effect:\n  --bfile genotyping_merged\n  --extract range genomic_regions.txt\n  --out snp_region_list_plink_results\n  --recode A\n\n15544 MB RAM detected; reserving 7772 MB for main workspace.\n129787 variants loaded from .bim file.\n488377 people (223429 males, 264742 females, 206 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to snp_region_list_plink_results.nosex .\n--extract range: 126452 variants excluded.\n--extract range: 3335 variants remaining.\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 488377 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.97396.\n3335 variants and 488377 people pass filters and QC.\nNote: No phenotypes present.\n--recode A to snp_region_list_plink_results.raw ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\nID                    file-GF2fzqjJ8v5bqvx94Kykyzkf\nClass                 file\nProject               project-G96qBfjJ8v5XFFfK7yxxx5y3\nFolder                /\nName                  snp_region_list_plink_results.raw\nState                 closing\nVisibility            visible\nTypes                 -\nProperties            -\nTags                  -\nOutgoing links        -\nCreated               Wed Jul  6 11:20:15 2022\nCreated by            lora.boteva\n via the job          job-GF2fX58J8v5bYY7Z2JV8QQpx\nLast modified         Wed Jul  6 11:20:26 2022\nMedia type            \narchivalState         \"live\"\ncloudAccount          \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}]}