{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Install plink\n","We recommend utilising the conda install of plink using the standard commands available here:\n","- https://anaconda.org/bioconda/plink\n","\n","Plink is also available in the following locations:\n","- https://www.cog-genomics.org/plink2/\n","- https://github.com/chrchang/plink-ng\n","\n","Once installed, continue with the below code chunks."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","plink --help | head"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","\n","# Use the code below to download files for chromosomes of interest\n","\n","\n","# list all available genotype calls files and places them in a file\n","# The file all_snp_files.txt will appear in your environment (click on the folder icon on the left of your screen to view files in your environment)\n","dx find data --name \"ukb22418_*.bed\" --path \"/Bulk/Genotype Results/Genotype calls/\" --delim ',' | awk -F, '{ print $4 }' |awk -F'/' '{print $5}' | awk -F'.' '{print $1}'  > all_snp_files.txt\n","\n","\n","# The code below provides an example on how to download data for  chromosome 1, chromosome 4 and chromosome X\n","# To modify, replace  the 1,4, and X in the followig expression '^ukb22418_c[14X]_' with your chromosome of interest\n","# For example, if you are interested in SNPs on chromosomes 5,6,8, and 9, use: '^ukb22418_c[5689]_'\n","# For double-digit chromosomes , e.g chromosome 12 use the following format: '^ukb22418_c[1][2]_'\n","# For mixed double digit and single digit chromosomes, e,g, chromosome 8 and 12, use the following format: '^ukb22418_c[18][2]_'\n","# To include all chromosomes modify the expression to '^ukb22418_c*_' \n","\n","\n","\n","files=$(grep '^ukb22418_c[14X]_' all_snp_files.txt) \n","echo $files \n","\n","\n","\n","\n","# Download files containing chromosomes of interest\n","\n","for i in $files; do\n","dx download \"/Bulk/Genotype Results/Genotype calls/${i}*\"\n","done\n","\n","\n","\n","# The downloaded files will appear in your environment. There should be a bed, bim and fam file for each chromosome containing SNPs of interest.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","\n","# Define files to merge. Please replace the '^ukb22418_c[14X]_' expression in the code below with the expression used to define chromosomes in the previous code chunk. \n","\n","grep '^ukb22418_c[14X]_' all_snp_files.txt > files_to_merge.txt\n","cat files_to_merge.txt\n","plink --merge-list files_to_merge.txt --make-bed --out genotyping_merged"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","# replace the rsIDs in the code below with your SNPs; Always seperate each rsID with a commma.\n","plink --bfile genotyping_merged --snps rs28659788,rs116587930 --recode A --out snp_ind_plink_results\n","dx upload snp_ind_plink_results.raw"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","\n","dx download -f \"snps_list.txt\"\n","plink --bfile genotyping_merged --extract snps_list.txt --recode A --out snp_list_plink_results\n","dx upload snp_list_plink_results.raw\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","plink --bfile genotyping_merged --chr 1 --from-kb 0 --to-kb 30000000 --recode A --out snp_region_plink_results\n","dx upload snp_region_plink_results.raw"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%bash\n","\n","dx download -f \"genomic_regions.txt\"\n","\n","plink --bfile genotyping_merged --extract range genomic_regions.txt  --recode A --out snp_region_list_plink_results\n","\n","dx upload snp_region_list_plink_results.raw\n"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.5"}},"nbformat":4,"nbformat_minor":4}
